### PR TITLE
Log with `logging` module in `qiwis`

### LINKED ISSUE #180: `Qiwis` central logging using `logging`
### Feature you want to implement
Use `logging` module for the `Qiwis` system's central logging.

### How the feature is implemented
Replace all the print-out parts with `logging` API.
Then they will appear depending on the logging configuration.

### Additional context
See also: https://docs.python.org/3.7/howto/logging.html#configuring-logging-for-a-library

### PR DESCRIPTION
**Pull request by kangz12345**
I added logging parts.
One can check the logging as follows:

Screenshot 2023-06-06 at 12 24 01

(The stack trace is printed by the logger, and the program is not dead, i.e., the exception is caught normally.)

If you are not familiar to `logging`, then I can make a test script for you.

This closes #180.

### CODE REVIEW

File: qiwis.py [hunk 1]
```diff
@@ -171,6 +178,7 @@ def load(self, appInfos: Mapping[str, AppInfo]):
         """
         for name, info in appInfos.items():
             self.createApp(name, info)
+        logger.info("Loaded %d app(s)", len(appInfos))
```
Comment (BECATRUE) [inline qiwis.py:181]: Do you prefer '% operator' to 'f-string'?
Indeed, I like the latter😅

File: qiwis.py [hunk 2]
```diff
@@ -266,6 +278,7 @@ def updateFrames(self, name: str):
             self.removeFrame(name, orgFrames[frame])
         for frame in newFramesSet - orgFramesSet:
             self.addFrame(name, frame, info)
+        logger.info("Updated frames: %d -> %d", len(orgFramesSet), len(newFramesSet))
```
Comment (BECATRUE) [inline qiwis.py:282]: The information looks not enough.
How about listing the frames of both `orgFrameSet` and `newFrameSet`?

File: qiwis.py [hunk 3]
```diff
@@ -287,7 +300,11 @@ def subscribe(self, app: str, channel: str):
             app: The name of the app which wants to subscribe to the channel.
             channel: The target channel name.
         """
-        self._subscribers[channel].add(app)
+        if app in self._subscribers[channel]:
```
Comment (BECATRUE) [inline qiwis.py:304]: 👍

File: qiwis.py [hunk 4]
```diff
@@ -303,7 +320,10 @@ def unsubscribe(self, app: str, channel: str) -> bool:
         try:
             subscribers.remove(app)
         except KeyError:
+            logger.error("The app %s tried to unsubscribe from %s, "
+                         "which it does not subscribe to", app, channel)
             return False
+        logger.info("The app %s unsubscribed from %s", app, channel)
```
Comment (BECATRUE) [inline qiwis.py:327]: Why don't you write it down in the present tense?
In `subscribe()`, you did that!

**Review by BECATRUE — CHANGES_REQUESTED** [review]
I left a few reviews!

File: qiwis.py (hunk 1 above)
Comment (kangz12345) [reply qiwis.py:181]: For other uses, I also prefer the 'f-string'!
However, the `logging` system does some [optimization](https://docs.python.org/3.7/howto/logging.html#optimization):
> Formatting of message arguments is deferred until it cannot be avoided.

If we use the f-string, it will be directly converted to a string, even if the message is not printed at all.

See also: [Logging Cookbook](https://docs.python.org/3.7/howto/logging-cookbook.html#using-particular-formatting-styles-throughout-your-application).
However, note that we cannot use other formatting options, e.g., `{}`-style, when we _construct_ each log message.

In short, we _have to_ use such `%`-style formatting if we want to exploit the optimization.

File: qiwis.py (hunk 2 above)
Comment (kangz12345) [reply qiwis.py:282]: Yes indeed.
I will include the class names of the frames.

File: qiwis.py (hunk 4 above)
Comment (kangz12345) [reply qiwis.py:327]: 😄 Nice observation, I think it is a matter of taste.
For me, the expression "subscribe to A" refers to a certain continued _status_, so since `subscribe()` is called, the app subscribes to the channel until it unsubscribes the channel.
However, "unsubscribing" is more like an event at a specific time point, so I used the past tense.

I think "subscribing" is also used as an event just like "unsubscribing", so it is totally fine to use the past tense for `subscribe()` for consistency!
Which one do you prefer?

File: qiwis.py (hunk 1 above)
Comment (BECATRUE) [reply qiwis.py:181]: Wow, I never thought it would be for this reason!
Now, never mind😃

File: qiwis.py (hunk 4 above)
Comment (BECATRUE) [reply qiwis.py:327]: I was convinced by the first explanation!
I think you can leave it as it is.

File: qiwis.py (hunk 2 above)
Comment (kangz12345) [reply qiwis.py:282]: Actually, I realized that `removeFrame()` and `addFrame()` will log the information instead.
So I updated `removeFrame()` to log the frame class name as well.

Comment (kangz12345) [reply qiwis.py:282]: Screen Shot 2023-06-08 at 19 07 33

Comment (BECATRUE) [reply qiwis.py:282]: Oh, you are right.
Then, the logging looks enough!

**Review by BECATRUE — APPROVED** [review]
LGTM👍

Then, please release this as the latest version!

**Comment by kangz12345** [comment]
> Then, please release this as the latest version!

Actually, there are several issues (#181, #182 and #184) left.
